### PR TITLE
test appveyor

### PR DESCRIPTION
**Pull request by simonbyrne**

### CODE REVIEW

**Comment by simonbyrne** [comment]
@tkelman I've enabled appveyor (https://ci.appveyor.com/project/simonbyrne/rmath-julia) but it doesn't seem to be trigerring. Is it because I don't have admin access to JuliaLang on github?

File: appveyor.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+shallow_clone: true
+# set clone depth
+clone_depth: 5
+
+init:
+# Carriage returns are bad
+  - git config --global core.autocrlf input
+
+build_script:
+  - set PATH="C:\mingw-w64\x86_64-6.3.0-posix-seh-rt_v5-rev1\bin"
```
Comment (tkelman) [inline appveyor.yml]: I doubt this will be ABI compatible

**Comment by tkelman** [comment]
likely need to manually set the webhook

File: appveyor.yml (hunk 1 above)
Comment (simonbyrne) [inline appveyor.yml]: isn't that how we build other windows libraries?

Comment (tkelman) [inline appveyor.yml]: we don't use this specific build. especially on 32 bit it's likely the wrong exception model

**Comment by simonbyrne** [comment]
> likely need to manually set the webhook

Can you do that?

File: appveyor.yml (hunk 1 above)
Comment (simonbyrne) [inline appveyor.yml]: okay, well could you give me a hand with this? or just create new binaries and upload them to the release until we have a more permanent solution in place?

**Comment by tkelman** [comment]
just need the webhook url, and to remember how it should be set

**Comment by nalimilan** [comment]
@simonbyrne You can find the webhook in the AppVeyor project settings. Just copy/paste it here.

File: appveyor.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+environment:
+  matrix:
+#  - ARCH: "i686"
+  - ARCH: "x86_64"
+
+# Only build on master and PR's for now, not personal branches
+# Whether or not PR's get built is determined in the webhook settings
+branches:
```
Comment (tkelman) [inline appveyor.yml]: this either needs to have some entries or not be included

**Comment by tkelman** [comment]
Simon sent it to me offline, should be enabled now

File: appveyor.yml (hunk 1 above)
Comment (ararslan) [inline appveyor.yml]: You might be able to do something like this: https://github.com/ararslan/fftw-builder/pull/4 (though that currently doesn't work, but something like that might work)

**Comment by tkelman** [comment]
msys2's gcc is definitely not abi compatible, especially on 32 bit

**Comment by simonbyrne** [comment]
> msys2's gcc is definitely not abi compatible, especially on 32 bit

What should I use for 32-bit then? the 64-bit one seems to work okay.

**Comment by simonbyrne** [comment]
@tkelman This seems to work. Any thoughts?

**Comment by tkelman** [comment]
use the cygwin cross compilers, not msys2. msys2 isn't supported by base julia, so packages shouldn't be using it either

**Comment by tkelman** [comment]
take msys2 off the path before testing anything, the gcc dlls linked to by an msys2 compiled dll aren't compatible with the ones used to build julia (they might be close enough on 64 bit, but on 32 bit it's a different exception handling model so libgcc has a different name)

**Comment by simonbyrne** [comment]
Okay, I've switched to mingw (for some reason I couldn't get it to work yesterday). Everything now seems fine.

File: appveyor.yml [hunk 3]
```diff
@@ -0,0 +1,60 @@
+environment:
+  matrix:
+  - ARCH: "i686"
+    CCPATH: 'C:\mingw-w64\i686-6.3.0-posix-dwarf-rt_v5-rev1\mingw32\bin'
```
Comment (tkelman) [inline appveyor.yml]: no this will not work.wrong exception handling model

Comment (simonbyrne) [inline appveyor.yml]: okay, well can you suggest which one I do use? The cygwin one didn't seem to work either.

Comment (simonbyrne) [inline appveyor.yml]: Should I be using the slij one? Is this documented anywhere?

**Comment by ararslan** [comment]
Can we put together a devdocs page that explain what people need to use for each platform in order to build a binary compatible with Julia? Apparently it's extraordinarily non-obvious.

**Comment by simonbyrne** [comment]
also which threading model should I be using?

File: appveyor.yml (hunk 3 above)
Comment (tkelman) [inline appveyor.yml]: what did you try on cygwin?

**Comment by tkelman** [comment]
it'll be most reliable to use the cygwin cross compiler, as that's what's used to build julia itself

File: appveyor.yml (hunk 3 above)
Comment (simonbyrne) [inline appveyor.yml]: https://github.com/JuliaLang/Rmath-julia/pull/19/commits/cb85628ba43763be70fdd96d6df70ff36ed12bae
which gave
https://ci.appveyor.com/project/simonbyrne/rmath-julia/build/1.0.33/job/kjus8eii39083v65

Comment (tkelman) [inline appveyor.yml]: using cygwin's gcc is only for compiling cygwin applications, which julia isn't. the cross compiler is x86_64-w64-mingw32-gcc

Comment (simonbyrne) [inline appveyor.yml]: which one should i use for 32bit?

**Comment by simonbyrne** [comment]
Okay, well this works now, and should be using the correct settings.

File: appveyor.yml (hunk 3 above)
Comment (tkelman) [inline appveyor.yml]: i686-w64-mingw32-gcc

**Comment by tkelman** [comment]
cygwin's cross compiler is using pthreads now i believe. the makefile changes here aren't really suitable for building in msys2 or cygwin. mingw32-make is really limited and generally shouldn't be used

**Comment by simonbyrne** [comment]
Which changes in particular?

**Comment by tkelman** [comment]
type and cmd. those make this harder to cross compile, and aren't necessary in either msys2 or cygwin. they're only necessary in a situation that no one ever does - wanting to run a makefile without having a posix emulation environment.

**Comment by ararslan** [comment]
> mingw32-make is really limited and generally shouldn't be used

What's the alternative on AppVeyor?

**Comment by simonbyrne** [comment]
Aren't they both built in windows commands?

**Comment by tkelman** [comment]
the alternative is normal full featured gnu make in cygwin or msys2. don't rely on anything that's pre installed on appveyor, since it's usually the wrong version and not going to be present on users' systems. I'd prefer if their vm's were basically empty, but that's a separate issue.

cross compiling doesn't always happen from windows.

**Comment by simonbyrne** [comment]
Okay, but how much does that matter? The Makefile isn't set up for any
multiple OS cross compile.

I just need a way to generate binaries for windows so I can fix this. If I
change the threads can we use these? (I don't think the library uses
threads if that matters).

**Comment by tkelman** [comment]
I don't think you should use mingw from chocolatey, no, as it's never been used for anything else with julia or tested beyond a minimal few ccalls here.

When this repo was a dependency of Base the cross compilation functionality was used often. Seems unnecessary to break it for compatibility with an unnecessary broken version of make that no one should ever use.

File: appveyor.yml [hunk 4]
```diff
@@ -31,14 +31,12 @@ init:
   - ps: "[System.Net.ServicePointManager]::SecurityProtocol = [System.Net.SecurityProtocolType]::Tls12"
 
 install:
-#  - appveyor DownloadFile https://cygwin.com/setup-x86_64.exe -FileName C:\projects\cygwin-setup-x86_64.exe
-#  - C:\projects\cygwin-setup-x86_64.exe -q -P make,p7zip,mingw64-i686-gcc-core,mingw64-x86_64-gcc-core
+  - appveyor DownloadFile https://cygwin.com/setup-x86_64.exe -FileName C:\projects\cygwin-setup-x86_64.exe
+  - C:\projects\cygwin-setup-x86_64.exe -q -D -P make,p7zip,mingw64-i686-gcc-core,mingw64-x86_64-gcc-core
+  - set PATH=c:\cygwin64\bin;c:\cygwin64\lib\p7zip
```
Comment (tkelman) [inline appveyor.yml]: don't do this, run a cygwin login shell with sh -l

Comment (simonbyrne) [inline appveyor.yml]: any particular reason why?

Comment (tkelman) [inline appveyor.yml]: cygwin is meant to handle its own environment, not be run piecemeal by modifying path externally

Comment (tkelman) [inline appveyor.yml]: make.exe isn't the entry point to run cygwin correctly, its shell is

Comment (simonbyrne) [inline appveyor.yml]: Is that documented anywhere: there are plenty of sites out there that suggest that this is okay. e.g. https://cygwin.com/ml/cygwin/2006-01/msg01476.html

Comment (simonbyrne) [inline appveyor.yml]: it seems that the shell that make uses is in fact cmd.exe, as unix paths for CC don't work

File: appveyor.yml [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+# Download and install most recent cygwin
+  - appveyor DownloadFile https://cygwin.com/setup-x86_64.exe -FileName C:\projects\cygwin-setup-x86_64.exe
+  - C:\projects\cygwin-setup-x86_64.exe -q -g -P make,mingw64-i686-gcc-core,mingw64-x86_64-gcc-core
+# Download most recent Julia Windows binary
+  - appveyor DownloadFile %JULIA_URL% -FileName C:\projects\julia-binary.exe
+# Set paths to only use this cygwin
+  - set PATH=C:\cygwin64\bin;"C:\Program Files\7-Zip"
+
+build_script:
+  - make CC=%ARCH%-w64-mingw32-gcc
```
Comment (simonbyrne) [inline appveyor.yml]: @tkelman it's way too complicated to figure out how to call make with the correct CC via cygwin sh, so unless you want to do it, I'm going to call it a day on this.

File: appveyor.yml (hunk 4 above)
Comment (tkelman) [inline appveyor.yml]: only because you're calling it in this way

File: appveyor.yml (hunk 5 above)
Comment (tkelman) [inline appveyor.yml]: `sh -lc "make -C /cygdrive/c/projects/Rmath-julia CC=$ARCH-w64-mingw32-gçc"`

Comment (simonbyrne) [inline appveyor.yml]: I'm assuming the cedilla there was a mistake...

Comment (tkelman) [inline appveyor.yml]: yes, where did that come from

**Comment by simonbyrne** [comment]
Okay, any objections to this?

**Comment by tkelman** [comment]
should be able to go back to using cygwin's p7zip now that this is running in the right shell, but otherwise looks fine now

**Pull request merged**